tf-dataflow/real_data_loader.py

```diff
@@ -0,0 +1,265 @@
+
+
+def _int64_feature(value):
+    """Returns an int64_list from a bool / enum / int / uint."""
+    return _int64_list_feature([value])
+
+
+def _int64_list_feature(value):
+    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))
+
+
+def _convert_to_example(filename, image_buffer, height, width, labels, bboxes):
+    colorspace = b"RGB"
+    channels = 3
+    image_format = b"JPG"
+    xmins, xmaxes, ymins, ymaxes = bboxes
+
+    example = tf.train.Example(
+        features=tf.train.Features(
+            feature={
+                "image/height": _int64_feature(height),
+                "image/width": _int64_feature(width),
+                "image/colorspace": _bytes_feature(colorspace),
+                "image/channels": _int64_feature(channels),
+                "image/object/class/label": _int64_list_feature(
+                    labels
+                ),  # model expects 1-based
+                "image/object/bbox/xmin": _float_list_feature(xmins),
+                "image/object/bbox/xmax": _float_list_feature(xmaxes),
+                "image/object/bbox/ymin": _float_list_feature(ymins),
+                "image/object/bbox/ymax": _float_list_feature(ymaxes),
+                "image/format": _bytes_feature(image_format),
+                "image/filename": _bytes_feature(
+                    bytes(os.path.basename(filename), "utf-8")
+                ),
+                "image/encoded": _bytes_feature(image_buffer),
+            }
+        )
+    )
+    return example
+
+
+def _get_image_data(filename, coder):
+    # Read the image file.
+    with tf.gfile.GFile(filename, "rb") as ifp:
+        image_data = ifp.read()
+
+    image = coder.decode_jpeg(image_data)
+
+    # Check that image converted to RGB
+    assert len(image.shape) == 3
+    height = image.shape[0]
+    width = image.shape[1]
+    assert image.shape[2] == 3
+
+    return image_data, height, width
+
+
+def _annotations_to_bb_normalized(values, width, height):
+    x_mins = [v["bbox"][0] / 5184.0 for v in values]
+    x_maxes = [x + v["bbox"][2] / 5184.0 for x, v in zip(x_mins, values)]
+    y_mins = [v["bbox"][1] / 3456.0 for v in values]
+    y_maxes = [y + v["bbox"][3] / 3456.0 for y, v in zip(y_mins, values)]
+    return x_mins, x_maxes, y_mins, y_maxes
+
+
+def convert_to_example(filename, annotations):
+    # ignore labels not in categories list
+    coder = ImageCoder()
+    image_buffer, height, width = _get_image_data(filename, coder)
+    del coder
+
+    # NOTE: Can break if non-bounding box annotation types are on
+    # TODO: Check annotations.json to find id of bbox annotations
+    labels = [v["label_id"] for v in annotations]
+    bboxes = _annotations_to_bb_normalized(annotations, width, height)
+    example = _convert_to_example(
+        filename, image_buffer, height, width, labels, bboxes
+    )
+    return example.SerializeToString()
+
+
+class ImageCoder(object):
```

Comment:
By default python3 class inherent from `object`. The syntex should be `class ImageCoder:`

Reply:
fixed